### ISSUE TITLE
Future Direction of Resque (Discussion Welcome)

### ISSUE THREAD

**Issue by chrisccerami** [resque/resque#1759]
Hey folks. As many of you are aware, Resque has had several phases of inactivity over the years. One of the biggest reasons has been that most of the people who were active maintainers have moved on to other things. It seems like there are still plenty of people who care about the future of Resque though, so I wanted to welcome some discussion on the topic. Also, if anyone is interested in becoming a member of the Resque team, please let us know how you'd like to get involved.

What is it you'd like to see from this library? Maintain the status quo and keep the lights on? A total rebuild from the ground up? In what ways can you help contribute to that future direction?

**Comment by iloveitaly** [comment]
For me, I'm interesting mostly in keeping the lights on and incremental improvements. Resque works well for my use case and a rebuild wouldn't really add value in my scenario.

**Comment by ajaya** [comment]
I would personally keep the lights on and take in some of the gems that are used to keep resque performing in production like https://github.com/stulentsev/resque-multi-job-forks etc.

**Comment by travisbell** [comment]
My thoughts exactly, @ajaya. It's not inherently a Resque (as in the singular gem) issue but rather the collective abandonment of many of the plugins that make Resque useful. I still quite happily run Resque in production but have ended up with a nasty mix of custom forks and unofficial gems to make the newer versions run.

If I was thinking about taking Resque to the next level, I would be considering bringing a lot of that functionality into the Resque core, giving us users more confidence that these features are properly supported and maintained.

**Comment by mockdeep** [comment]
I'm game. I've put a bunch of work into the RMagick gem, along with [a couple of other great contributors](https://github.com/rmagick/rmagick/graphs/contributors). I don't have a ton of time for actual coding these days, but I can almost always make time for code review and responding to new issues. I'm very much a code gardener type, focused on maintenance and code quality.

**Comment by 64kramsystem** [comment]
I'm also for maintenance and small features. A rewrite would be very expensive, in particular, if there isn't a stable team.

Adding sponsorship may be a beneficial idea, in order to motivate (not "pay" :grin:) stable maintainers. After all, the only players in the database-agnostic area are Resque and Sidekiq (DelayedJob is unmaintaned, as far as I see) - and Sidekiq can be expensive, so I think there is still space for Resque.

What I'd like to see: distributed scheduler; this actually forced use to move to Sidekiq.

**Comment by iloveitaly** [comment]
@64kramsystem could you explain the distributed scheduler use case further? Curious why this was a requirement for you.

**Comment by 64kramsystem** [comment]
> @64kramsystem could you explain the distributed scheduler use case further? Curious why this was a requirement for you.

In short: to have a fleet of hosts which uniformly act as both web server and queue processor.

Now... uh. I had a look of resque-scheduler (which we use/used), and when we had a look around (mid-year), it hadn't been maintained for an year, and hadn't any release for two. But shortly after, unbeknownst to us, development resumed :sweat_smile:.

Based on a quick read, if I understand correctly, `resque-scheduler` does fit the bill as of now (I'm not sure if I should be happy or not :sweat_smile:).

Thanks for the attention, and for the developiment :smile:

**Comment by nevans** [comment]
I strongly agree that a more "batteries included" approach to the most commonly used plugins would be useful.

If there's interest, I'll (eventually, haha) submit a PR to merge nevans/resque-pool (or something like it) into core. Resque-pool still sees a *lot* of downloads, but I think resqued has seen more active maintenance?

**Comment by iloveitaly** [comment]
@nevans I think an easy improvement here is to to improve discoverability.

* There's no website right now
* Lots of information is in the wiki, which is hard to search and doesn't seem to be indexed effectively by google
* There's a huge readme. It's hard to parse and omits lots of important information that's in the wiki
* There's not a good list of up to date plugins. For instance, I think the old sentry plugin is still referenced while sentry has since built in resque support directly in their client libraries.
* Popular plugins like resque-pool, resque-retry, etc are not hosted on the resque org which makes them harder to find.

Keeping things more modular feels like the right approach for now, esp while there isn't a ton of community involvement in development right now. If we can improve some of the above documentation issues, and get more community development involvement, I think it could make sense to merge more plugins into master.

**Comment by nvolker** [comment]
I'd like to get involved 🖐️

**Comment by iloveitaly** [comment]
@nvolker amazing! Couple places you could get started:

* Any issue marked with 'help wanted' https://github.com/resque/resque/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22
* Specifically, I'd love to update http://resque.github.io and generally reorganize/update the documentation. If you have time for this, it would be great to make some progress. I left notes in my previous comment about what we could do clean this up (migrate the wiki to the resque.github.io site would be the first step).

**Comment by 64kramsystem** [comment]
@iloveitaly Would you add the GitHub option to sponsor this and the other related projects¹? I think it's a significant aspect for established projects. Moreover, for some use cases, the main competitor has an... infinitely higher price :wink:

¹=the scheduler, please!!

**Comment by Anas198900** [comment]
[]()

**Comment by xhs345** [comment]
Agree with what most are saying. No rebuild needed, but it would be nice to integrate some gems that are useful and in some cases abandoned at this point. For example we are using:

- resque-scheduler
- resque-job-stats (had to fork for compatibility)
- resque-retry 
- resque-lock-timeout (had to fork for compatibility)

Would also love to see resque-pool

**Comment by joshburnsxyz** [comment]
I second @xhs345 (and the majority census by the looks) and would like to see some 3rd party gems integrated and become "standard". Would love to get involved

**Comment by iloveitaly** [comment]
@joshburnsxyz @xhs345 I think the first step would be moving them all to the resque org to at least make maintenance easier. Could either of you contact the owners and see if they are interested in migrating them to our org?

**Comment by joshburnsxyz** [comment]
@iloveitaly I can send out a few emails later tonight

**Comment by joshburnsxyz** [comment]
@iloveitaly @xhs345
Did we have an "official list" of candidates or just the ones mentioned on this thread?

**Comment by iloveitaly** [comment]
@joshburnsxyz We don't have an official list. I would like to see as many of the popular plugins managed centrally, and then we can slowly discuss moving some of them into master.

I've been chatting with the resque-retry folks and I'll be moving that over soon.

**Comment by varyonic** [comment]
We use:
- resque-scheduler
- resque-retry
- resque-cleaner
- resque-throttler
- resque-lonely_job

**Comment by iloveitaly** [comment]
@varyonic I'm open to pulling as many of them into the resque org as possible, and then we can think about how to best merge the most popular into the core resque library.

**Comment by nvolker** [comment]
Hey, there! Another question that comes to mind for me is... Does Resque have an alternative to https://github.com/breamware/sidekiq-batch? I've seen:
- https://github.com/annkissam/resque-batch
- https://github.com/drfeelngood/resque-batched-job

But have concerns on how maintained and supported they are? We have strong cases for doing batch jobs with all the hooks (bells and whistles) ready to go. I could build something custom, but I figured there must be something the community is using rather than converting over to SideKiq.

**Comment by iloveitaly** [comment]
I haven't used these plugins at all, and they don't seem that popular. I can't speak to how well they work.

That being said, if there's support for putting more maintenance/development effort behind them, I'm in support of bringing them into the resque org to give them more visibility.

**Comment by pboling** [comment]
I maintain a suite of resque extension gems, and I'd like to join the resque organization, and potentially transfer the suite to the organization.

I rewrote, and renamed, both `resque_solo` and `resque-lonely_job`, because they
 can't be used together, but since one offers run time uniqueness, and the other queue time uniqueness, I needed both.  Why?  Their `redis_key` methods directly conflict, among other more subtle conflicts.

@wallace, author of `resque-lonely_job`, transferred control of that library to me, and I plan to create a migration path from [`resque-lonely_job`](https://github.com/wallace/resque-lonely_job) to [`resque-unique_at_runtime`](https://github.com/pboling/resque-unique_at_runtime). 

The following allow complete control over handling of uniqueness for jobs:

- [resque-unique_by_arity](https://github.com/pboling/resque-unique_by_arity) - Allows fine-tuned handling of arity (argument composition) comparisons for determination of job uniqueness
- [`resque-unique_at_runtime`](https://github.com/pboling/resque-unique_at_runtime) - Job run time uniqueness 
- [`resque-unique_in_queue`](https://github.com/pboling/resque-unique_in_queue) - Job queue time uniqueness

**Comment by dhaulakhandi** [comment]
Rescue should have the ability to send email notification when some queue is > SOME_LIMITATION so monitoring is easier?

**Comment by nhooda** [comment]
I'm interested in keeping the lights on and incremental improvements.

**Comment by chrishough** [comment]
Just catching up on this, could be interested in helping out too, wasn't sure what the status was. With Sidekiq's expensive pricing model it would be great to have a solid alternative in the marketplace

**Comment by iloveitaly** [comment]
PRs are welcome! There's lots of PRs that need a little bit of additional
effort to get over the finish line, so feel free to jump in on those as
well.

**Comment by pboling** [comment]
@iloveitaly Did you see my comment above about my suite of resque uniqueness plugins, and have you considered moving them to the `resque` org?  I'd be happy to continue maintaining them, even though I don't currently have a use case for Resque (also one could appear at any moment!).

**Comment by iloveitaly** [comment]
@pboling I didn't! If you are open it, let's migrate them over. You should be able to go to the `Settings` area of each repo and transfer ownership to the `@resque` org. Thanks for being open to migrating them over! They seem like great additions to the resque ecosystem.

**Comment by QuentinLemCode** [comment]
We use Resque in production on a big website, with ~500 workers and ~6M jobs processed by day.
The gem `resque-scheduler` is a must have and, like some comments said, this should be included into the Resque lib.
Others thing we do and would be great to include : 
- Batch jobs (with limit on parallel execution)
- Rate limiting
- Asynchronous Pub/Sub
- Circuit breaker (put jobs in a delayed queue when an API is down)

**Comment by QuentinLemCode** [comment]
Another proposal would be to rely on Redis PubSub to avoid polling from the workers

**Comment by pboling** [comment]
@iloveitaly I'd love to transfer the repos to the resque org, but I'll have to be added to the org first, as it is a requirement of the transfer (I just tried it and it failed because I'm not).
<img width="755" alt="Screenshot 2023-08-15 at 3 04 14 PM" src="https://github.com/resque/resque/assets/19505/ae64136e-0bd5-4cc9-8858-18d11913e905">

**Comment by pboling** [comment]
@corincerami @iloveitaly ping!

**Comment by iloveitaly** [comment]
@pboling just invited you!

**Comment by PatrickTulskie** [comment]
Random but I'm at GitHub universe. If anyone wants to chat IRL about resque stuff, lemme know.

**Comment by pboling** [comment]
@iloveitaly I have transferred the repos, but now I can no longer manage them.  I'm no longer a maintainer of them, or one an "owners" team, I suppose.  Can you make it so I can help with them again?  I'd like to help with a few other things around here too!

**Comment by corincerami** [comment]
@pboling I've promoted you to an owner of the Resque org on Github
